win32ss/printing/providers/localspl/printerdrivers.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * PROJECT:     ReactOS Local Spooler
+ * LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * PURPOSE:     Functions for printer driver information
+ * COPYRIGHT:   Copyright 2018 Mark Jansen (mark.jansen@reactos.org)
+ */
+
+#include "precomp.h"
+
+
+// Local Constants
+static DWORD dwDriverInfo1Offsets[] = {
+    FIELD_OFFSET(DRIVER_INFO_1W, pName),
+    MAXDWORD
+};
+
+static DWORD dwDriverInfo2Offsets[] = {
+    FIELD_OFFSET(DRIVER_INFO_2W, pName),
+    FIELD_OFFSET(DRIVER_INFO_2W, pEnvironment),
+    FIELD_OFFSET(DRIVER_INFO_2W, pDriverPath),
+    FIELD_OFFSET(DRIVER_INFO_2W, pDataFile),
+    FIELD_OFFSET(DRIVER_INFO_2W, pConfigFile),
+    MAXDWORD
+};
+
+static DWORD dwDriverInfo3Offsets[] = {
+    FIELD_OFFSET(DRIVER_INFO_3W, pName),
+    FIELD_OFFSET(DRIVER_INFO_3W, pEnvironment),
+    FIELD_OFFSET(DRIVER_INFO_3W, pDriverPath),
+    FIELD_OFFSET(DRIVER_INFO_3W, pDataFile),
+    FIELD_OFFSET(DRIVER_INFO_3W, pConfigFile),
+    FIELD_OFFSET(DRIVER_INFO_3W, pHelpFile),
+    FIELD_OFFSET(DRIVER_INFO_3W, pDependentFiles) | MARSHAL_MULTI_SZ,
+    FIELD_OFFSET(DRIVER_INFO_3W, pMonitorName),
+    FIELD_OFFSET(DRIVER_INFO_3W, pDefaultDataType),
+    MAXDWORD
+};
+
+static void
+_LocalGetPrinterDriverLevel1(PLOCAL_PRINTER_HANDLE pHandle, PDRIVER_INFO_1W* ppDriverInfo, PBYTE* ppDriverInfoEnd, PDWORD pcbNeeded)
+{
+    DWORD n;
+    PWSTR pwszStrings[1];
+
+    pwszStrings[0] = pHandle->pPrinter->pwszPrinterDriver;
+
+    // Calculate the string lengths.
+    if (!ppDriverInfo)
+    {
+        for (n = 0; n < _countof(pwszStrings); ++n)
+        {
+            *pcbNeeded += (wcslen(pwszStrings[n]) + 1) * sizeof(WCHAR);
+        }
+
+        *pcbNeeded += sizeof(DRIVER_INFO_1W);
+        return;
+    }
+
+
+    // Finally copy the structure and advance to the next one in the output buffer.
+    *ppDriverInfoEnd = PackStrings(pwszStrings, (PBYTE)(*ppDriverInfo), dwDriverInfo1Offsets, *ppDriverInfoEnd);
+    (*ppDriverInfo)++;
+}
+
+static void
+_LocalGetPrinterDriverLevel2(PLOCAL_PRINTER_HANDLE pHandle, PDRIVER_INFO_2W* ppDriverInfo, PBYTE* ppDriverInfoEnd, PDWORD pcbNeeded)
+{
+    DWORD n;
+    PWSTR pwszStrings[5];
+
+    pwszStrings[0] = pHandle->pPrinter->pwszPrinterDriver;  // pName
+    pwszStrings[1] = L"Win32 x86";  // pEnvironment
+    pwszStrings[2] = L"c:\\reactos\\system32\\localspl.dll";  // pDriverPath
```

Comment:
Is this just for passing the WINE Test or does the implementation fulfill any real purpose?

localspl.dll is a Print Provider, not a Printer Driver. Apart from this, I probably don't need to tell you about the hardcoded path.

Reply:
MSDN didnt document this field as being optional, so I just dropped something in there.
I don't know of anything using it, but figured returning something rather than nothing is less likely to break stuff.